library/modules/EventManager.cpp

```diff
@@ -711,29 +742,32 @@ static void manageBuildingEvent(color_ostream& out) {
             continue;
         }
         buildings.insert(a);
-        for (auto &key_value : copy) {
-            EventHandler &handle = key_value.second;
-            DEBUG(log,out).print("calling handler for created building event\n");
-            handle.eventHandler(out, (void*)intptr_t(a));
-        }
+        new_buildings.emplace_back(a);
+
     }
     nextBuilding = *df::global::building_next_id;
 
+    std::for_each(new_buildings.begin(), new_buildings.end(), [&](int32_t building){
+        for (auto &[_,handle] : copy) {
+            DEBUG(log,out).print("calling handler for created building event\n");
+            handle.eventHandler(out, (void*)intptr_t(building));
+        }
+    });
+
     //now alert people about destroyed buildings
```

Comment:
could you take this opportunity to move the destroyed buildings events before the new building events? this would solve a longstanding bug in our building location cache where a new building is added in the spot a removed building just vacated.

Reply:
done!

Reply:
thanks! I keep on forgetting to do that. it's been bugging me for more than a year. It wasn't causing errors, just unnecessary O(N) lookups.